pygam/pygam.py

```diff
@@ -1870,6 +1876,260 @@ def gridsearch(self, X, y, weights=None, return_scores=False,
         else:
             return self
 
+    def sample(self, X, y, quantity='y', sample_at_X=None,
+               weights=None, n_draws=100, n_bootstraps=1, objective='auto'):
+        """Simulate from the posterior of the coefficients and smoothing params.
+
+        Samples are drawn from the posterior of the coefficients and smoothing
+        parameters given the response in an approximate way. The GAM must
+        already be fitted before calling this method; if the model has not
+        been fitted, then an exception is raised. Moreover, it is recommended
+        that the model and its hyperparameters be chosen with `gridsearch`
+        (with the parameter `keep_best=True`) before calling `sample`, so that
+        the result of that gridsearch can be used to generate useful response
+        data and so that the model's coefficients (and their covariance matrix)
+        can be used as the first bootstrap sample.
+
+        These samples are drawn as follows. Details are in the reference below.
+
+        1. `n_bootstraps` many "bootstrap samples" of the response (`y`) are
+        simulated by drawing random samples from the model's distribution
+        evaluated at the expected values (`mu`) for each sample in `X`.
+        2. A copy of the model is fitted to each of those bootstrap samples of
+        the response. The result is an approximation of the distribution over
+        the smoothing parameter `lam` given the response data `y`.
+        3. Samples of the coefficients are simulated from a multivariate normal
+        using the bootstrap samples of the coefficients and their covariance
+        matrices.
+
+        NOTE: A `gridsearch` is done `n_bootstraps` many times, so keep
+        `n_bootstraps` small. Make `n_bootstraps < n_draws` to take advantage
+        of the expensive bootstrap samples of the smoothing parameters.
+
+        NOTE: For now, the grid of `lam` values is the default of `gridsearch`.
+        Until randomized grid search is implemented, it is not worth setting
+        `n_bootstraps` to a value greater than one because the smoothing
+        parameters will be identical in each bootstrap sample.
+
+        Parameters
+        -----------
+        X : array of shape (n_samples, m_features)
+              empirical input data
+
+        y : array of shape (n_samples,)
+              empirical response vector
+
+        quantity : {'y', 'coef', 'mu'}, default: 'y'
+            What quantity to return pseudorandom samples of.
+            If `sample_at_X` is not None and `quantity` is either `'y'` or
+            `'mu'`, then samples are drawn at the values of `X` specified in
+            `sample_at_X`.
+
+        sample_at_X : array of shape (n_samples_to_simulate, m_features) or
+        None, default: None
+            Input data at which to draw new samples.
+
+            Only applies for `quantity == 'y'`. If `None`, then `sample_at_X`
```

Comment:
also applies if `quantity == 'mu'` right?

Reply:
Yep. I'll change that.

Reply:
Fixed